Hyper/MarchingCubes/ChunkWorker.cs

```diff
@@ -0,0 +1,114 @@
+        private ConcurrentDictionary<Guid, Chunk> _existingChunks;
+        private CancellationTokenSource _cancellationTokenSource;
+        private ScalarFieldGenerator _scalarFieldGenerator;
+        private ChunkFactory _chunkFactory;
+        private SimulationManager<NarrowPhaseCallbacks, PoseIntegratorCallbacks> _simulationManager;
+
+        public ChunkWorker(ConcurrentDictionary<Guid, Chunk> existingChunks, ScalarFieldGenerator scalarFieldGenerator, 
+            SimulationManager<NarrowPhaseCallbacks, PoseIntegratorCallbacks> simulationManager)
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+            _neededChunkPositions = new ConcurrentDictionary<Vector3i, Guid>();
+            _neededChunkPositions.TryAdd(new Vector3i(0, 0, 0), Guid.NewGuid()); // initial chunk...
+            _existingChunkPositions = new ConcurrentDictionary<Vector3i, Guid>();
+            _existingChunks = existingChunks;
+            _scalarFieldGenerator = scalarFieldGenerator;
+            _chunkFactory = new ChunkFactory(_scalarFieldGenerator);
+            _simulationManager = simulationManager;
+        }
+        public void StartProcessing()
+        {
+            var processingThread = new Thread(ProcessItems);
+            processingThread.IsBackground = true;
+            processingThread.Start();
+        }
+
+        public void StopProcessing()
+        {
+            _cancellationTokenSource.Cancel();
+        }
+
+        private void ProcessItems()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                Parallel.ForEach(_neededChunkPositions, kvp =>
+                {
+                    if (!_existingChunkPositions.ContainsKey(kvp.Key))
+                    {
+                        _existingChunkPositions.TryAdd(kvp.Key, kvp.Value);
+                        var chunk = _chunkFactory.GenerateChunk(kvp.Key);
+                        chunk.CreateCollisionSurface(_simulationManager.Simulation, _simulationManager.BufferPool);
+                        _existingChunks.TryAdd(kvp.Value, chunk);
+                    }
+                });
+                Parallel.ForEach(_existingChunkPositions, kvp =>
+                {
+                    if (!_neededChunkPositions.ContainsKey(kvp.Key))
+                    {
+                        _existingChunkPositions.TryRemove(kvp.Key, out _);
+                        _existingChunks.TryRemove(kvp.Value, out var chunk);
+                        // multithreaded removal might need a mutex?
+                        // chunk?.DisposeCollisionSurface(_simulationManager.Simulation, _simulationManager.BufferPool);
+                    }
+                });
+                Thread.Sleep(100);
+            }
+        }
+
+        public static void UpdateNeededChunksBasedOnPosition(ChunkWorker chunkWorker, Vector3 position)
+        {
+            const int chunkCreateRadius = 50;
```

Comment:
IF i change these to 150 and 170 there is a problem. Then i cross a chunk boarder I get a lag spike for about 0.5 seconds. I think its bacaouse ChunkWorker eats all the resources. Maybe limit the number of threads or something so that the loading takes a little bit longer but we do not get lag spikes

Reply:
fixed with a semaphore

Reply:
the issue still exists but only on debug mode